### PR TITLE
post node details of newly discovered node

### PR DESCRIPTION
**Pull request by sunnyqianzhang**
1. Add nodeId, macAddress and ipAddress to a newly discovered node
2. Use webhook to post this information to predefined website
@pengz1 @iceiilin

### CODE REVIEW

**Comment by iceiilin** [comment]
looks good to me. add @yyscamper @anhou to check whether it is reasonable to use lookup table here

File: lib/jobs/node-discovered-alert-job.js [hunk 1]
```diff
@@ -43,16 +45,22 @@ function nodeAlertJobFactory(
      */
     NodeAlertJob.prototype._run = function run() {
         var self = this;
+        var nodeIp = lookupService.nodeIdToIpAddresses(self.nodeId);
```
Comment (anhou) [inline lib/jobs/node-discovered-alert-job.js]: the return value is a promise. it's easily confused, if it's not in a promise chain

File: lib/jobs/node-discovered-alert-job.js [hunk 2]
```diff
@@ -43,16 +45,22 @@ function nodeAlertJobFactory(
...
         return waterline.nodes.needByIdentifier(self.nodeId)
         .then(function(node) {
             assert.string(node.type);
 
             if(self.context.data){
-                return eventsProtocol.publishNodeEvent(node, 'discovered', self.context.data);
-            }
-            else{
-                return eventsProtocol.publishNodeEvent(node, 'discovered');
+                nodeInfo.contextData = self.context.data;
```
Comment (anhou) [inline lib/jobs/node-discovered-alert-job.js]: I'm not sure the context.data details, if it's proper to put mac and ip to context.data to decrease nested values.

File: lib/jobs/node-discovered-alert-job.js (hunk 1 above)
Comment (yyscamper) [inline lib/jobs/node-discovered-alert-job.js]: If the property value is promise, you could try `Promise.props`

File: lib/jobs/node-discovered-alert-job.js [hunk 3]
```diff
@@ -43,16 +45,22 @@ function nodeAlertJobFactory(
      */
     NodeAlertJob.prototype._run = function run() {
         var self = this;
+        var nodeIp = lookupService.nodeIdToIpAddresses(self.nodeId);
+        var nodeMac = lookupService.ipAddressToMacAddress(nodeIp);
+        var nodeInfo = {
+                          'ipAddress': nodeIp,
+                          'macAddress': nodeMac,
+ 			   'nodeId': self.nodeId
```
Comment (yyscamper) [inline lib/jobs/node-discovered-alert-job.js]: indent problem

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 4]
```diff
@@ -47,6 +49,8 @@ describe("Job.Alert.Node.Discovered", function () {
         it("should _run() pass without additional data", function() {
             this.sandbox.stub(waterline.nodes, 'needByIdentifier').resolves({ type: 'compute' });
             this.sandbox.stub(eventsProtocol, 'publishNodeEvent').resolves();
+            this.sandbox.stub(lookupService, 'nodeIdToIpAddresses').returns("asdfagagafgdgdgdg");
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: `nodeIdToIpAddress` is designed to return a promise, so you need to stub it return promise as well.

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 5]
```diff
@@ -56,12 +60,16 @@ describe("Job.Alert.Node.Discovered", function () {
                     { type: 'compute' },
                     'discovered'
                 );
+               expect(lookupService.nodeIdToIpAddresses).to.have.been.calledOnce;
+		expect(lookupService.ipAddressToMacAddress).to.have.been.calledOnce;
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: indent problem

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 6]
```diff
@@ -72,13 +80,22 @@ describe("Job.Alert.Node.Discovered", function () {
                 expect(eventsProtocol.publishNodeEvent).to.have.been.calledWith(
                     { type: 'compute' },
                     'discovered',
-                    job.context.data
+		    {
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: indent error

**Comment by sunnyqianzhang** [comment]
@yyscamper @anhou Double checked the code and made modifications as the ipAddressToMacAddress and nodeIdToIpAddresses did return promise. Bigger changes are made, so I used push -f directly. 

@yyscamper Promise.pop is not suitable for my case since  nodeIdToIpAddresses rely on the result of ipAddressToMacAddress 

@yyscamper self.context data basically returns an object, but  in my book, put it in a separate field in parallel with node details make the structure more clear, so I left it as it is.

**Comment by sunnyqianzhang** [comment]
jenkins: depends on RackHD/on-core#258

**Comment by sunnyqianzhang** [comment]
jenkins: depends on https://github.com/RackHD/on-core/pull/258

**Comment by changev** [comment]
test this please

**Comment by changev** [comment]
test this please

**Comment by sunnyqianzhang** [comment]
jenkins: depends on https://github.com/RackHD/on-core/pull/258

**Comment by sunnyqianzhang** [comment]
test this please

File: lib/jobs/node-discovered-alert-job.js [hunk 7]
```diff
@@ -43,16 +45,19 @@ function nodeAlertJobFactory(
...
-            else{
-                return eventsProtocol.publishNodeEvent(node, 'discovered');
+            return lookup.findIpMacAddresses(self.nodeId)
+            .then(function(IpMacPairs) {
+               nodeInfo.IpMacAddress = IpMacPairs;
+               nodeInfo.nodeId = self.nodeId;
+               if(self.context.data){
+                  nodeInfo.contextData = self.context.data;
             }
+	    return eventsProtocol.publishNodeEvent(node, 'discovered', nodeInfo);
```
Comment (yyscamper) [inline lib/jobs/node-discovered-alert-job.js]: indent problem

File: lib/jobs/node-discovered-alert-job.js [hunk 8]
```diff
@@ -43,16 +45,19 @@ function nodeAlertJobFactory(
...
-                return eventsProtocol.publishNodeEvent(node, 'discovered', self.context.data);
-            }
-            else{
-                return eventsProtocol.publishNodeEvent(node, 'discovered');
+            return lookup.findIpMacAddresses(self.nodeId)
+            .then(function(IpMacPairs) {
+               nodeInfo.IpMacAddress = IpMacPairs;
+               nodeInfo.nodeId = self.nodeId;
+               if(self.context.data){
+                  nodeInfo.contextData = self.context.data;
```
Comment (yyscamper) [inline lib/jobs/node-discovered-alert-job.js]: The attribute name `contextData` is not suitable in Node Event, what kind of data is stored in the `context.data`? I think we could pick all nesting attribltes (in context.data) to the upper level (the same level with `nodeId` and `IpMacAddress`)

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 9]
```diff
@@ -47,21 +49,35 @@ describe("Job.Alert.Node.Discovered", function () {
         it("should _run() pass without additional data", function() {
             this.sandbox.stub(waterline.nodes, 'needByIdentifier').resolves({ type: 'compute' });
             this.sandbox.stub(eventsProtocol, 'publishNodeEvent').resolves();
+	    this.sandbox.stub(lookupService, 'findIpMacAddresses').resolves([
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: indent problem. I think you need to check the editor setting, sees your PRs usually have a lot indent problems.

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 10]
```diff
@@ -47,21 +49,35 @@ describe("Job.Alert.Node.Discovered", function () {
...
 
             return job._run()
             .then(function () {
                 expect(waterline.nodes.needByIdentifier).to.have.been.calledOnce;
                 expect(eventsProtocol.publishNodeEvent).to.have.been.calledOnce;
                 expect(eventsProtocol.publishNodeEvent).to.have.been.calledWith(
                     { type: 'compute' },
-                    'discovered'
+                    'discovered',
+		    {
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: Still a indent problem here.

File: spec/lib/jobs/node-discovered-alert-job-spec.js [hunk 11]
```diff
@@ -72,13 +88,21 @@ describe("Job.Alert.Node.Discovered", function () {
                 expect(eventsProtocol.publishNodeEvent).to.have.been.calledWith(
                     { type: 'compute' },
                     'discovered',
-                    job.context.data
+                    {
+                        contextData: { something: "passed in" },
+                        nodeId: "bc7dab7e8fb7d6abf8e7d6ab",
+                        IpMacAddress: [
+			{ipAddress: '1.1.1.1', macAddress: "aa:bb:cc:dd" },
```
Comment (yyscamper) [inline spec/lib/jobs/node-discovered-alert-job-spec.js]: Still indent problem

**Comment by changev** [comment]
test this please

**Comment by sunnyqianzhang** [comment]
test this please

File: lib/jobs/node-discovered-alert-job.js [hunk 12]
```diff
@@ -46,13 +48,19 @@ function nodeAlertJobFactory(
...
-                return eventsProtocol.publishNodeEvent(node, 'discovered');
-            }
+            return lookup.findIpMacAddresses(self.nodeId)
+            .then(function(ipMacPairs) {
+                var nodeInfo = {};
+                nodeInfo.ipMacAddresses = ipMacPairs;
+                nodeInfo.nodeId = self.nodeId;
+                if(self.context.data){
+                    nodeInfo = _.defaults(
+                        _.pick(nodeInfo, ['ipMacAddresses', 'nodeId']),
```
Comment (pengz1) [inline lib/jobs/node-discovered-alert-job.js]: _.pick is not necessary to me.

**Pull request merged**